proposals/11-21-krm-fn-k8s-backend.md

```diff
@@ -0,0 +1,218 @@
+on `docker` and makes `kubectl` self-sufficient to run krm-function based plugins.
+
+## Motivation
+
+1. https://github.com/GoogleContainerTools/kpt/issues/2158
+1. https://github.com/kubernetes-sigs/kustomize/pull/4260
+
+
+**Goals:**
+
+1. To enrich container-based plugin-extension implementation by introducing the new backend that
+will be an alternative to `docker` and that will allow to run krm-functions inside k8s using its native mechanisms to run containers
+1. To use k8s-native mechanisms to control access of krm-function to the environment it is running in
+
+
+**Non-goals:**
+
+1. Change Krm-functions spec
+1. Add new fields/settings to container-based kustomize plugins-extensions
+
+## Proposal
+
+The basis of this proposal is to give ability for the user to select that the krm-function would be ran
+*as-a-pod* rather than using `docker` (which will still be a default behavior that will guarantee no braking changes).
+For that purpose, in addition to the existing `--enable-exec`, `--enable-starlark` flags that have a very similar
+purpose - to allow the user *explicitly* say that it's ok to run krm-function not only in docker, we propose to add
+`--enable-as-a-pod` flag that will switch backend from `docker` to `kubectl`.
+
+In case the user has several k8s clusters, it's necessary to give him the ability to select
+where the krm-container will run. Kubectl has a pretty big list of parameters that allow
+to set kubeconfig location, context, namespace and etc. Adding all of them would be troublesome, but
+since we're only passing all of them to kubectl it's much easier to introduce only one parameter,
+`--pod-kubectl-args` that will contain all that arguments as one string, e.g.
+`--context kind-kustomize-api-test -n special-namespace`. This will give the user the maximum
+flexibility without limiting him. Even if kubectl will add some new parameters in future they all automatically
+will be supported.
+
+The current `docker` based implementation runs container in as-hermetic-as-possible environment:
+
+1. only env variables specified in function config OR env variables explicitly listed by the user are passed to krm-container.
+1. if krm-function requires network access it can be enabled only by explicitly set `--network` flag.
+1. if krm-function needs some additional mounted file/directory this also can be done only via setting a special `--mount` flag explicitly.
+
+That means that all non-default non-hermetic features except env variables are restricted and can be set only explicitly.
+Similarly, the restrictions applied to the k8s pod has to be set *explicitly* by the user with the same exception of env variables.
+
+Unfortunately, it's not possible to completely block network for a pod - networking is a k8s crucial part. But if needed
+it's possible to apply NetworkPolicies to the pod that match to it by labels. Since NetworkPolicy is a standard method
```

Comment:
IMO we should make running-in-pods as secure as running-docker-locally. Network access should not be allowed by default. 
We provide a guideline about how to setup the `NetworkPolicies` in the cluster.

Reply:
I've added the note about guide to setup `NetworkPolicies`. See lines 77-80.